pkgs/dart_tooling_mcp_server/lib/src/mixins/analyzer.dart

```diff
@@ -0,0 +1,182 @@
+  /// stop listening to their streams instead, so we store the stream
+  /// subscriptions instead of the watchers themselves.
+  final List<StreamSubscription> _watchSubscriptions = [];
+
+  @override
+  FutureOr<InitializeResult> initialize(InitializeRequest request) {
+    if (request.capabilities.roots == null) {
+      throw StateError(
+          'This server requires the "roots" capability to be implemented.');
+    }
+    registerTool(analyzeFilesTool, _analyzeFiles);
+    initialized.then(_listenForRoots);
+    return super.initialize(request);
+  }
+
+  @override
+  Future<void> shutdown() async {
+    await super.shutdown();
+    await _analysisContexts?.dispose();
+    for (var subscription in _watchSubscriptions) {
+      unawaited(subscription.cancel());
+    }
+    _watchSubscriptions.clear();
+  }
+
+  /// Lists the roots, and listens for changes to them.
+  ///
+  /// Whenever new roots are found, creates a new [AnalysisContextCollection].
+  void _listenForRoots([void _]) async {
+    rootsListChanged!.listen((event) async {
+      unawaited(_analysisContexts?.dispose());
+      _analysisContexts = null;
+      _createAnalysisContext(await listRoots(ListRootsRequest()));
+    });
+    _createAnalysisContext(await listRoots(ListRootsRequest()));
+  }
+
+  /// Creates an analysis context from a list of roots.
+  //
+  // TODO: Better configuration for the DART_SDK location.
+  void _createAnalysisContext(ListRootsResult result) async {
+    final sdkPath = Platform.environment['DART_SDK'];
+    if (sdkPath == null) {
+      throw StateError('DART_SDK environment variable not set');
+    }
+
+    final paths = <String>[];
+    for (var root in result.roots) {
+      var uri = Uri.parse(root.uri);
+      if (uri.scheme != 'file') {
+        throw ArgumentError.value(
+            root.uri, 'uri', 'Only file scheme uris are allowed for roots');
+      }
+      paths.add(p.normalize(uri.path));
+    }
+
+    for (var subscription in _watchSubscriptions) {
+      unawaited(subscription.cancel());
+    }
+    _watchSubscriptions.clear();
+
+    for (var rootPath in paths) {
+      var watcher = DirectoryWatcher(rootPath);
+      _watchSubscriptions.add(watcher.events.listen((event) {
+        final context = _analysisContexts?.contextFor(event.path);
+        if (context == null) return;
+        context.changeFile(p.normalize(event.path));
+      }));
+    }
+
+    _analysisContexts =
+        AnalysisContextCollection(includedPaths: paths, sdkPath: sdkPath);
+  }
+
+  /// Implementation of the [analyzeFilesTool], analyzes the requested files
+  /// under the requested project roots.
+  Future<CallToolResult> _analyzeFiles(CallToolRequest request) async {
+    var contexts = _analysisContexts;
+    if (contexts == null) {
+      return CallToolResult(content: [
+        TextContent(
+            text: 'Analysis not yet ready, please wait a few seconds and try '
+                'again.')
+      ], isError: true);
```

Comment:
Can we use and enforce the new Dart formatter on this codebase?

Reply:
Previously the flutter stable we were getting didn't support it but it looks like it should now so I can try upping our min SDK again to do that. Will do it as a separate CL to reduce churn here.

Reply:
(filed https://github.com/dart-lang/ai/issues/46)